pageserver/src/object_repository.rs

```diff
@@ -327,87 +392,114 @@ impl Timeline for ObjectTimeline {
     ///
     /// This will implicitly extend the relation, if the page is beyond the
     /// current end-of-file.
-    fn put_wal_record(&self, tag: BufferTag, rec: WALRecord) -> Result<()> {
+    fn put_wal_record(&self, tag: ObjectTag, rec: WALRecord) -> Result<()> {
         let lsn = rec.lsn;
 
         self.put_page_entry(&tag, lsn, PageEntry::WALRecord(rec))?;
-        debug!(
-            "put_wal_record rel {} blk {} at {}",
-            tag.rel, tag.blknum, lsn
-        );
+        debug!("put_wal_record {:?} at {}", tag, lsn);
 
-        // Also check if this created or extended the file
-        let old_nblocks = self.relsize_get_nowait(tag.rel, lsn)?.unwrap_or(0);
+        if let ObjectTag::RelationBuffer(tag) = tag {
+            // Also check if this created or extended the file
+            let old_nblocks = self.relsize_get_nowait(tag.rel, lsn)?.unwrap_or(0);
 
-        if tag.blknum >= old_nblocks {
-            let new_nblocks = tag.blknum + 1;
+            if tag.blknum >= old_nblocks {
+                let new_nblocks = tag.blknum + 1;
 
-            trace!(
-                "Extended relation {} from {} to {} blocks at {}",
-                tag.rel,
-                old_nblocks,
-                new_nblocks,
-                lsn
-            );
+                trace!(
+                    "Extended relation {} from {} to {} blocks at {}",
+                    tag.rel,
+                    old_nblocks,
+                    new_nblocks,
+                    lsn
+                );
 
-            self.put_relsize_entry(&tag.rel, lsn, RelationSizeEntry::Size(new_nblocks))?;
-            let mut rel_meta = self.rel_meta.write().unwrap();
-            rel_meta.insert(
-                tag.rel,
-                RelMetadata {
-                    size: new_nblocks,
-                    last_updated: lsn,
-                },
-            );
+                self.put_relsize_entry(&tag.rel, lsn, RelationSizeEntry::Size(new_nblocks))?;
+                let mut rel_meta = self.rel_meta.write().unwrap();
+                rel_meta.insert(
+                    tag.rel,
+                    RelMetadata {
+                        size: new_nblocks,
+                        last_updated: lsn,
+                    },
+                );
+            }
         }
-
         Ok(())
     }
 
-    /// Unlink object. This method is used for marking dropped relations.
+    /// Unlink relation. This method is used for marking dropped relations.
     fn put_unlink(&self, rel_tag: RelTag, lsn: Lsn) -> Result<()> {
```

Comment:
Is it not possible to unlink other kinds of objects? How are prepared transactions removed from `pg_twophase` when they commit/abort?

Reply:
No other objects are unlinked through object repository API.
SLRU pages are truncated, but there is special method for them.
Two-phase records are deleted by GC when them are out of GC (PITR) horizon and are completed (committed or aborted). So we do not special method in object repository API to remove them.

Reply:
>  SLRU pages are truncated, but there is special method for them.

Could we use unlink for SLRU segments too, and avoid the special method?

> Two-phase records are deleted by GC when them are out of GC (PITR) horizon and are completed (committed or aborted). So we do not special method in object repository API to remove them.

That doesn't sound right. You need to remove the pg_twophase file as soon as the transaction commits or aborts. Please add a test case for that scenario:

1. Start up the system. Prepare a transaction at LSN 100
2. Run more queries, run COMMIT PREPARED at LSN 150
3. Stop and recreate the compute node
4. Check if the transaction is committed

Reply:
> Could we use unlink for SLRU segments too, and avoid the special method?
Originally it was really singe put_unlink method for both cases. But now you put unlink tag inside RelationSizeEntry.
So it can not be used any more for non-relational objects. 
And actually I think that it having separate method for SLRU is really better, because semantic is very different. In case of relation we really drop all relation and in case of SLRU we just truncate it.

Reply:
>That doesn't sound right. You need to remove the pg_twophase file as soon as the transaction commits or aborts.
It is not true. Transaction may be committed/aborted but if we restore database at the point preceding this event, it should be still in progress. 
So 2PC objects should be removed by GC as any other objects - only when them are beyond GC horizon.
But there is also extra check for them: if transaction is still in progress, then we can not remove it even if it is beyond horizon.

Reply:
> Please add a test case for that scenario:

There is test_twophase.py test. I do not understand what are the differences of the proposed scenario from this test.

Reply:
Ah, you're filtering out the twophase entries of committed transactions when you create the tarball, in `add_twophase_file`. That almost works. The problem I see is that there is no guarantee that the clog page is still around when `add_twophase_file` is called. The sequence would be something like this:

1. Transaction with id 100 is prepared and committed
2. A lot more transactions happen
3. Postgres runs vacuum and truncates away the clog segment containing XID 100
4. Page server receives the clog truncation request
5. A new Postgres node is initialized, and a basebackup tarball is created. The twophase object is still present in the repository, because garbage collection hasn't run. `add_twophase_file` finds it, and calls `get_tx_status` for XID 100. The call fails, because the clog page has already been truncated away.

Reply:
My assumption was that we are not supporting two-phase transactions older than PITR interval (gc_interval).
Please notice that our object store is versioned and Postgres VACUUM do not have complete control on it.
Yes, it can truncate CLOG but correspondent pages will be still present in object storage until them cross GC horizon.

Unfortunately it is not checked now. The only solution I can propose at this moment, is that GC should cacluclate min(xid) of all prepared transactions and use it as low margin for CLOG truncation.

Reply:
I wonder whether it is too utopian idea not to truncate CLOG at all? It can not be larger than 8Gb. Is 8Gb per Postgres instance too large space overhead for our storage?

Reply:
I think that instead of calculating min(xid) for all prepred xacts, we can calculate min(lsn) and set GC horizon for SLRUs as min(last_valid_lsn - pitr_interval, prepared_horizon).
It is based on the assumption that any commits aborts of prepared transaction have LSN > prepare_record.lsn
If prepared transaction is not yet completed, then preserving CLOG for its XID is enforced by Postgres. And if is commited or aborted then shifting gc_horizon to prepare_record.lsn should preserve correspondent CLOG page.

Do you agree?

Reply:
Or you could call `put_unlink` for the twophase state file, at the commit/abort record. I'm scared by treating some object kinds specially.

Reply:
> > Could we use unlink for SLRU segments too, and avoid the special method?
> 
> Originally it was really singe put_unlink method for both cases. But now you put unlink tag inside RelationSizeEntry.
> So it can not be used any more for non-relational objects.
> And actually I think that it having separate method for SLRU is really better, because semantic is very different. In case of relation we really drop all relation and in case of SLRU we just truncate it.

Do you think of the whole CLOG (or any other SLRU) as one object that's stored in the repositry, or do you think of each segment, i.e. each physical clog file in Postgres, as a separate object? If you think of each segment as a separate object, then SLRU truncation really is the same as unlinking the older objects, and you don't need to treat SLRU truncation as a special case in object_repository.rs.

Reply:
Yes, I agree with this point of view. And this is actually why originally there was single put_unlink method.
But as  I already mentioned, it is not possible now when Unlink is madepart of  RelationSizeEntry.

Reply:
> Or you could call `put_unlink` for the twophase state file, at the commit/abort record. I'm scared by treating some object kinds specially.

Yes, it is also possible. But approach with pinning CLOG seems to be more correct. If we want to perform PITR at some point where prepared transaction was in progress, we need to also preserve correspondent CLOG pages.